### DISCUSSION TITLE
Import cells from Hoc definitions to python Cell class?

### DISCUSSION THREAD

**Discussion by jtajchman** [jonescompneurolab/hnn-core#706] [Q&A] [answered]
I'm trying to model external electric fields (for TMS) in a cortical network using HNN-Core. Because a neuron's response to E-fields depends on its morphology, particularly that of the axon, I want to import realistic neuron models despite the increased computational cost. I am also using a custom module to add E-fields to the simulations.

There are a set of cells I would like to use ([ModelDB](https://modeldb.science/241165)), though they are defined in Hoc, not in the way that HNN likes. These are associated with Aberra et. al. ([2018](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6239949/)) and subsequent studies in ([2020](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6889021/)) and ([2023](https://doi.org/10.1016/j.neuroimage.2023.120184)). They describe cells specifically designed for simulation of TMS response. 

My specific question is whether I could take these cells defined in Hoc-NEURON and use them in HNN simulation. As opposed to the normal HNN workflow of defining the cell in a Python dictionary-based format and then using that to construct the cell in NEURON, is there a way to do the opposite? How could I create the Python dictionaries from the Hoc files/define the cells in HNN's native format, allowing for full compatibility? Or could I use the Hoc definitions directly in simulation/bypass the dictionary step, but possibly lose some functionality?

I believe this question is somewhat similar to the issue of merging HNN-Core's definition of a network to that of NetPyNE, which I've read is an active work in progress. If there are any solutions or resources to help with the problem, I would appreciate it. 
Thank you for reading!

**Comment by rythorpe** [comment, marked as answer]
Hi @jtajchman, unfortunately this is currently not possible in HNN-core due to how our `Network` class interacts with individual neurons. Specifically, it relies on hard-coded assumptions about the number, type, and name of specific neuron compartments.

As you noted, we're currently working on refactoring the code so that new neuron types and network structures can be loaded into HNN-core more easily, though there is much work to do. If you're interested in helping, we always welcome new contributors! :smile:

**Reply by jtajchman** [reply]
Thank you so much! It's helpful to know where effort is worth putting in a project. Where should I look further if I wanted to contribute to this aspect of HNN-Core?

**Reply by rythorpe** [reply]
We're currently working on updating our I/O methods for writing (reading) a `Network` object to (from) disk, but then we want to implement 1 of 2 options: create an API for reading in arbitrary neuron types compiled in NEURON or read in a network specification file compiled in NETPYNE.  We're honestly still in the process of trying to decide which route is better, particularly since we're not a network building tool (we leave that for tools like NETPYNE). IDK if that answers your question, but also feel free to checkout our [contributing guide](https://github.com/jonescompneurolab/hnn-core/blob/master/doc/contributing.rst).
